relayer/src/chain/cosmos/batch.rs

```diff
@@ -139,27 +157,41 @@ async fn send_messages_as_batches(
 }
 
 fn batch_messages(
+    config: &TxConfig,
     max_msg_num: MaxMsgNum,
     max_tx_size: MaxTxSize,
+    key_entry: &KeyEntry,
+    account: &Account,
+    tx_memo: &Memo,
```

Comment:
This parameter lists is used repetitively, maybe there should be a struct to arrange some of them.

Reply:
It was originally not repetitive, but you made them repetitive now. The point of explicit parameters is to make it obvious of cascading dependencies and help us avoid that.

Instead, why don't just pass `tx_envelope_len` as a parameter and call `encoded_tx_len` from the call site? It also looks like `encoded_tx_len` can be estimated once and applicable for all transactions, rather than recomputing it every time.

Reply:
@soareschen The same parameters are used in other functions in this module, I did not want to break the pattern at this point. As you point out, this flattening was done in an earlier refactoring to reduce unnecessary data dependencies and useless mutability, but I think we should follow it by some creative parameter grouping where it makes sense, to avoid excessive repetition.

> It also looks like `encoded_tx_len` can be estimated once and applicable for all transactions

I was not so sure, there may be some variation in how e.g. addresses are encoded, and there's also the memo field. There needs to be some context above the `send_batched_messages_and_wait_*` API to safely make this assumption.

Reply:
The key challenge for using a common struct in this case is that the references are borrowed from different places. So if we define a struct like `TxContext` we would have to move or clone the values into it. And if we define a struct containing references like `TxContext<'a>`, then it does not really contribute much other than being a syntactic sugar. It would also risk overgeneralizing functions, especially when catering for those few places that requires a `&mut Account`.

That said, if we want to refactor the code, it would better to define a proper `TxHandler` struct that specializes in sending transactions, and decouple it from the `ChainHandle` implementation, which is the real source of this mess.

The address should all have the same length, and is the same within the same `ChainHandle`. Same for the memo field, which has the same value configured. The only thing that could potentially change length is the account sequence, unless it has a fixed length encoding like `u64`.

Reply:
> t would better to define a proper `TxHandler` struct that specializes in sending transactions, and decouple it from the `ChainHandle` implementation, which is the real source of this mess.

I agree, splitting the "god trait" into smaller traits is a good way to refactor.

> Same for the memo field, which has the same value configured.

So there might be a case for a struct that contains these less-variable parameters.

> The only thing that could potentially change length is the account sequence, unless it has a fixed length encoding like u64.

Yes, there is a number of these varint fields and length delimiters that can grow between one tx to another, so you either have to compute the length on the fully composed transaction as it would be sent, or put in some upper bound for the overhead, neither of which is an ideal approach.